src/main/java/org/opensearch/knn/index/query/BaseQueryFactory.java

```diff
@@ -90,20 +100,71 @@ protected static Query getFilterQuery(BaseQueryFactory.CreateQueryRequest create
                 createQueryRequest.getFieldName()
             )
         );
+
+        // preserve nestedStack
+        Deque<ObjectMapper> nestedLevelStack = new LinkedList<>();
+        ObjectMapper objectMapper = null;
+        while ((objectMapper = queryShardContext.nestedScope().getObjectMapper()) != null) {
+            nestedLevelStack.push(objectMapper);
+            queryShardContext.nestedScope().previousLevel();
+        }
+
         final Query filterQuery;
         try {
             filterQuery = createQueryRequest.getFilter().get().toQuery(queryShardContext);
         } catch (IOException e) {
             throw new RuntimeException("Cannot create query with filter", e);
+        } finally {
+            while ((objectMapper = nestedLevelStack.peek()) != null) {
+                queryShardContext.nestedScope().nextLevel(objectMapper);
+                nestedLevelStack.pop();
```

Comment:
Can we call this instead of peek in while loop?

Reply:
same as https://github.com/opensearch-project/k-NN/pull/2641#discussion_r2039669066